File: C1-Algorithmic_Toolbox/W2-Algorithmic_Warmup/2_last_digit_of_fibonacci_number/fibonacci_last_digit.py
# python3


def fibonacci_last_digit(n):
    # Naive solution:
    if n <= 1:
        return n
    previous = 0
    current = 1
    for _ in range(n - 1):
        previous, current = current, previous + current
    return current % 10


def fibonacci_last_digit_fast(n):
    # Faster solution: Avoid adding two large numbers because it's SLOW.
    if n <= 1:
        return n
    previous = 0
    current = 1
    for _ in range(2, n + 1):
        res = (previous + current) % 10
        previous = current
        current = res
    return current % 10


if __name__ == "__main__":
    n = int(input())
    print(fibonacci_last_digit_fast(n))

File: C1-Algorithmic_Toolbox/W2-Algorithmic_Warmup/2_last_digit_of_fibonacci_number/fibonacci_last_digit.cpp
#include <iostream>

using namespace std;

int get_fibonacci_last_digit_naive(int n) {
    if (n <= 1) return n;

    int previous = 0;
    int current = 1;
    int res = 0;

    for (int i = 2; i <= n; ++i) {
        res = previous + current;
        previous = current;
        current = res;
    }

    return current % 10;
}

int get_fibonacci_last_digit_fast(int n) {
    if (n <= 1) return n;

    int previous = 0;
    int current = 1;
    int res = 0;

    for (int i = 2; i <= n; ++i) {
        res = (previous + current) % 10;
        previous = current;
        current = res;
    }

    return current % 10;
}

// g++ -pipe -O2 -std=c++14 fibonacci_last_digit.cpp -lm && ./a.out
int main() {
    int n;
    cin >> n;
    int c = get_fibonacci_last_digit_fast(n);
    cout << c << '\n';
}
